util/conditions/v1beta2/getter.go

```diff
@@ -0,0 +1,206 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1beta2
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+// TODO: Move to the API package.
+const (
+	// NoReasonReported identifies a clusterv1.Condition that reports no reason.
+	NoReasonReported = "NoReasonReported"
+)
+
+// Getter interface defines methods that an API object should implement in order to
+// use the conditions package for getting conditions.
+type Getter interface {
+	// GetV1Beta2Conditions returns the list of conditions for a cluster API object.
+	// Note: GetV1Beta2Conditions will be renamed to GetConditions in a later stage of the transition to V1Beta2.
+	GetV1Beta2Conditions() []metav1.Condition
+}
+
+// Get returns a condition from the object implementing the Getter interface.
+//
+// Please note that Get does not support reading conditions from unstructured objects nor from API types not implementing
+// the Getter interface. Eventually, users can implement wrappers on those types implementing this interface and
+// taking care of aligning the condition format if necessary.
+func Get(sourceObj Getter, sourceConditionType string) *metav1.Condition {
+	// if obj is nil, the requested condition does not exist.
+	if sourceObj == nil {
+		return nil
+	}
+
+	// if obj is a zero value of its type (a pointer to nil), the requested condition does not exist.
+	if ptr := reflect.ValueOf(sourceObj); ptr.Kind() == reflect.Pointer {
+		elem := ptr.Elem()
+		if !elem.IsValid() {
+			return nil
+		}
+	}
+
+	// Otherwise get the requested condition.
+	return meta.FindStatusCondition(sourceObj.GetV1Beta2Conditions(), sourceConditionType)
+}
+
+// UnstructuredGet returns a condition from an Unstructured object.
+//
+// UnstructuredGet supports retrieving conditions from objects at different stages of the transition from
+// clusterv1.conditions to the metav1.Condition type:
+//   - Objects with clusterv1.Conditions in status.conditions; in this case a best effort conversion
+//     to metav1.Condition is performed, just enough to allow surfacing a condition from a provider object with Mirror
+//   - Objects with metav1.Condition in status.v1beta2.conditions
+//   - Objects with metav1.Condition in status.conditions
+func UnstructuredGet(sourceObj runtime.Unstructured, sourceConditionType string) (*metav1.Condition, error) {
+	// if obj is nil, the requested condition does not exist.
+	if sourceObj == nil {
+		return nil, nil
+	}
+
+	// if obj is a zero value of its type (a pointer to nil), the requested condition does not exist.
+	if ptr := reflect.ValueOf(sourceObj); ptr.Kind() == reflect.Pointer {
+		elem := ptr.Elem()
+		if !elem.IsValid() {
+			return nil, errors.Errorf("source object is not a valid unstructured type: %v", sourceObj)
```

Comment:
nit: I think we shouldn't use %v to print the entire sourceObj

Reply:
fixed